File: Graph/547. Number of Provinces/547.java
class Solution {
    public void go(int city, int[][] isConnected, boolean[] visited){
        int N = isConnected[city].length;
        for(int i = 0; i < N; i++){
            if(i != city && !visited[i] && isConnected[city][i]==1){
                visited[i] = true;
                go(i, isConnected, visited);
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        int N = isConnected.length;
        boolean[] visited = new boolean[N];
        int provinces = 0;

        for(int i = 0; i < N; i++){
            if(!visited[i]){
                provinces ++;
                visited[i] = true;
                go(i, isConnected, visited);
            }
        }

        return provinces;
    }
}

File: Graph/DFS/547. Number of Provinces/547.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        provinces = 0
        visited = [False] * len(isConnected)
        N = len(isConnected)

        def go(city):
            for nei, connected in enumerate(isConnected[city]):
                if nei != city and not visited[nei] and connected:
                    visited[nei] = True
                    go(nei)

        for city in range(N):
            if not visited[city]:
                provinces += 1
                visited[city] = True
                go(city)

        return provinces
